Add a Smart Contracts ABI page for LSP2Utils.sol, LSP5Utils.sol, LSP6Utils.sol & LSP10Utils.sol (#414)
lukso-network/docs: docs/standards/smart-contracts/utils/lsp2-json-schema-utils.md

---
title: LSP2JSONSchemaUtils
sidebar_position: 1
---

# LSP2JSONSchemaUtils

:::info

[`LSP2Utils.sol`](https://github.com/lukso-network/lsp-smart-contracts/blob/develop/contracts/LSP2ERC725YJSONSchema/LSP2Utils.sol)

:::

This library can be used to generate ERC725Y keys according to the JSON schemas defined in the LSP2 standard.

## Functions

### generateSingletonKey

```solidity
function generateSingletonKey(
    string memory keyName
) internal pure returns (bytes32);
```

Generates a data key of `keyType` Singleton.

#### Parameters:

| Name      | Type   | Description                                          |
| :-------- | :----- | :--------------------------------------------------- |
| `keyName` | string | The string to hash to generate a Singleton data key. |

#### Return Values:

| Name     | Type    | Description                      |
| :------- | :------ | :------------------------------- |
| `result` | bytes32 | Data key of `keyType` Singleton. |

### generateArrayKey

```solidity
function generateArrayKey(
    string memory keyName
) internal pure returns (bytes32);
```

Generates a data key of `keyType` Array by hashing `keyName`.

##### Requirements:

> - The last two characters should be `[]`.

Reviewer: ```suggestion
- The last two characters MUST be `[]`.
```

E.g.:

```solidity
string memory keyName = "ArrayName[]";
```

#### Parameters:

| Name      | Type   | Description                                                            |
| :-------- | :----- | :--------------------------------------------------------------------- |
| `keyName` | string | The string that will be used to generate a data key of `keyType` Array |

#### Return Values:

| Name     | Type    | Description                  |
| :------- | :------ | :--------------------------- |
| `result` | bytes32 | Data key of `keyType` Array. |

### generateArrayElementKeyAtIndex

```solidity
function generateArrayElementKeyAtIndex(
    bytes32 arrayKey,
    uint256 index
) internal pure returns (bytes32);
```

> Generates an Array index data key by concatenating the first 16 bytes of `arrayKey` with an `index`.

Reviewer 2: Generates the element key of the array at the given index , something like this ..

> Generates the element key of the array at the given index by concatenating the first 16 bytes of `arrayKey` with an `index`.

Reviewer: ```suggestion
Generates the element data key for the `arrayKey` at the given `index`. This is done by concatenating the first 16 bytes of `arrayKey` with an `index`.
```

#### Parameters:

| Name      | Type    | Description                                                                       |
| :-------- | :------ | :-------------------------------------------------------------------------------- |

> | `keyName` | string  | The key from which we're getting the first half of the Array index data key from. |

Reviewer 2: I would simply say the array data key

> | `index`   | uint256 | Used to generate the second half of the Array index data key.                     |

Reviewer 2: also here, simple saying the index of the element In the array

#### Return Values:

| Name     | Type    | Description                        |
| :------- | :------ | :--------------------------------- |

> | `result` | bytes32 | Data key of `keyType` Array Index. |

Reviewer 2: Maybe the data key of the element in the array ?

| :----- | :---- | :-------------------------------- |
| `data` | bytes | The value that is to be verified. |

### isBytes4EncodedArray

```solidity
function isBytes4EncodedArray(
    bytes memory data
) internal pure returns (bool);
```

Verify that `data` is an array of bytes4 (bytes4[]) encoded according to the Solidity ABI specs.

#### Parameters:

| Name   | Type  | Description                       |
| :----- | :---- | :-------------------------------- |
| `data` | bytes | The value that is to be verified. |

### isCompactBytesArray

```solidity
function isCompactBytesArray(
    bytes memory compactBytesArray
) internal pure returns (bool);
```

Verify the validity of the `compactBytesArray` according to LSP2.

#### Parameters:

| Name                | Type  | Description                       |
| :------------------ | :---- | :-------------------------------- |
| `compactBytesArray` | bytes | The value that is to be verified. |

### uncheckedIncrement

```solidity
uncheckedIncrement(
    uint256 i
) internal pure returns (uint256);
```

Will return unchecked incremented uint256.
Can be used to save gas when iterating over loops.

## References

> - [Solidity implementations (GitHub)](https://github.com/lukso-network/lsp-universalprofile-smart-contracts/tree/develop/contracts)

Reviewer 2: Not really related to your changes, but this link is quite old, we can replace it with the new one but wonder if we should keep it or remove it
